docs/OBJECT_LOCK_TEST_PLAN.md

```diff
@@ -0,0 +1,157 @@
+- putting invalid object lock configuration returns error
+    - ObjectLockEnabled !== "Enabled"
+    - Rule object doesn't contain DefaultRetention key
+    - Mode !== "GOVERNANCE" or "COMPLIANCE"
+    - Days are not an integer
+    - Years are not an integer
+
+### getObjectLockConfiguration tests
+
+- getting configuration as non-bucket-owner user returns AccessDenied error
+- getting configuration when none is set returns
+ObjectLockConfigurationNotFoundError error
+- getting configuration returns correct object lock configuration for bucket
+
+### putObjectRetention
+
+- putting retention as non-bucket-owner user returns AccessDenied error
+- putting retention on object in bucket without object lock enabled returns
+InvalidRequest error
+- putting valid retention period updates object metadata
+
+### getObjectRetention
+
+- getting retention as non-bucket-owner user returns AccessDenied error
+- getting retention when none is set returns NoSuchObjectLockConfiguration
+error
+- getting retention returns correct object retention period
+
+### putObjectLegalHold
+
+- putting legal hold as non-bucket-owner user returns AccessDenied error
+- putting legal hold on object in bucket without object lock enabled returns
+InvalidRequest error
+- putting valid legal hold updates object metadata
+
+### getObjectLegalHold
+
+- getting legal hold as non-bucket-owner user returns AccessDenied error
+- getting legal hold when none is set returns NoSuchObjectLockConfiguration
+error
+- getting legal hold returns correct object legal hold
+
+## End to End Tests
+
+### Scenarios
+
+- Create bucket with object lock enabled. Put object. Put object lock
+configuration. Put another object.
+
+    - Ensure object put before configuration does not have retention period set
+    - Ensure object put after configuration does have retention period set
+
+- Create bucket without object lock. Put object. Enable object lock with token
+and put object lock configuration. Put another object.
+
+    - Ensure object put before configuration does not have retention period set
+    - Ensure object put after configuration does have retention period set
+
+- Create bucket with object lock enabled and put configuration with COMPLIANCE
+mode. Put object.
+
+    - Ensure object cannot be deleted (returns AccessDenied error).
+    - Ensure object cannot be overwritten.
+
+- Create bucket with object lock enabled and put configuration with GOVERNANCE
+mode. Put object.
+
+    - Ensure user without permission cannot delete object
+    - Ensure user without permission cannot overwrite object
+    - Ensure user with permission can delete object
+    - Ensure user with permission can overwrite object
+    - Ensure user with permission can lengthen retention period
+    - Ensure user with permission cannot shorten retention period
+
+- Create bucket with object lock enabled and put configuration with date past
+expiration. Put object.
```

Comment:
I think this is another case (putting configuration with date past expiration) which we don't make possible in the first place if I'm not mistaken.